File: Programmers/Level 1/가장 큰 수.js
// 틀림
function solution(numbers) {
    var strArr = numbers.map(number => number.toString())
    var maxLeng = strArr.map(char => char.length).reduce((a, b) => Math.max(a, b))
    var sortedArr = strArr.map(function(number) {
        var idx = strArr.indexOf(number)
        if (number.length < maxLeng) {
            for (let i=0; i<maxLeng-number.length+1; i++) {
                number += number[0];
            }
        }
        return [number, idx]
    })
    sortedArr = sortedArr.sort((a, b) => b[0] - a[0])
    var answer = [];
    for (let j=0; j<sortedArr.length; j++) {
        answer.push(strArr[sortedArr[j][1]])
    }
    if (answer[0] == 0) {
        return '0'
    } else return answer.join('')
}


// 조건 하나 빼먹은듯..
function solution2(numbers) {
    const maxLeng = 4
    const unsortedObj = {};
    for (let i=0; i<numbers.length; i++) {
        let stringifiedNum = numbers[i].toString()
        stringifiedNum += stringifiedNum[0].repeat(maxLeng-stringifiedNum.length+1)
        unsortedObj[stringifiedNum] = numbers[i]
    }
    const sortedObj = Object.values(unsortedObj).reverse()
    return sortedObj.join('')
}


const sortFunc = array => {
    if (array.length < 2) return array;

    const pivot = array[array.length - 1];
    const left = [], right = [];

    for (let i = 0; i < array.length - 1 ; i++) {
        let compare = array[i]
        if (compare[0] > pivot[0]) {
            left.push(compare)
        } else if (compare[0] === pivot[0]) {
            if (compare.length == pivot.length) {
                if (compare > pivot) {
                    left.push(compare)
                } else right.push(compare);
            } else {
                const cLength = compare.length
                const pLength = pivot.length
                if (cLength > pLength) {
                    const pivot2 = pivot + pivot[0].repeat(cLength - pLength)
                    if (compare > pivot2) {
                    left.push(compare)
                    } else if (compare == pivot2) {
                        if (compare + pivot > pivot + compare) {
                            left.push(compare)
                        } else right.push(compare)
                    } else right.push(compare);
                } else if (cLength < pLength) {
                    const compare2 = compare + compare[0].repeat(pLength - cLength)
                    if (compare2 > pivot) {
                        left.push(compare)
                    } else if (compare2 === pivot) {
                        if (compare + pivot > pivot + compare) {
                            left.push(compare)
                        } else right.push(compare)
                    } else right.push(compare);
                }
            }
        }
        else right.push(compare);
    }
    return [...sortFunc(left), pivot, ...sortFunc(right)];
};

function solution3(numbers) {
    var strArr = numbers.map(number => number.toString())
    var answer = sortFunc(strArr)
    if (answer[0] == 0) {
        return '0'
    } else return answer.join('')
}

// 이게 찐 정답이었다..
function solution4(numbers) {
    const strArr = numbers.map((number) => number.toString())
    const sortedNumbers = strArr.sort((a, b) => (b + a) - (a + b))
    return sortedNumbers[0] === '0' ? '0' : sortedNumbers.join('')
}

File: Programmers/Level 2/가장 큰 수.py
# 틀림
def solution(numbers):
	str_arr = list(map(str, numbers))
	max_length = 0
	for num in str_arr:
		if len(num) > max_length:
			max_length = len(num)
	sorted_arr = []
	for idx, num in enumerate(str_arr):
		if len(num) < max_length:
			num += num[0] * (max_length - len(num))
		sorted_arr += [[int(num), idx]]
	sorted_arr.sort(key=lambda x: -x[0])
	answer = []
	for num in sorted_arr:
		answer += str_arr[num[1]]
	return '0' if answer[0] == '0' else ''.join(answer)


# 몇 개 테스트케이스가 시간초과
def compare(a, b):
	if int(b + a) - int(a + b) > 0:
		return True
	else:
		return False

def solution2(numbers):
	str_arr = list(map(str, numbers))
	str_arr.sort(key=lambda x:x[0], reverse=True)
	flag = 0
	while flag != len(numbers) - 1:
		flag = 0
		for i in range(len(numbers) - 1):
			if not compare(str_arr[i], str_arr[i+1]):
				flag += 1
			else:
				str_arr[i], str_arr[i+1] = str_arr[i+1], str_arr[i]
	return '0' if str_arr[0] == '0' else ''.join(str_arr)


# 람다와 cmp_to_key 를 좀 더 봐야겠다 싶은 코드..
# cpm_to_key 는 고차 함수를 반환한다.
def solution3(numbers):
	import functools
	answer = list(map(str, numbers))
	answer.sort(key=functools.cmp_to_key(lambda a, b: int(b+a)-int(a+b)))
	return '0' if answer[0] == '0' else ''.join(answer)
